File: yetutil/src/main/java/net/yet/ui/widget/recycler/XActionRecyclerView.java
package net.yet.ui.widget.recycler;

import android.content.Context;
import android.graphics.Color;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import net.yet.theme.Colors;
import net.yet.ui.util.XView;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * Created by yet on 2015/10/28.
 */
public abstract class XActionRecyclerView<T> extends RecyclerView {
    private final List<T> items = new ArrayList<>();
    private int normalColor = Color.WHITE;
    private int pressColor = Colors.Fade;
    private boolean enableItemBackColor = true;
    private int actionColor = Colors.TRANS;
    private int actionPress = Colors.Fade;
    private boolean actionColorEnable = true;
    private LinearLayoutManager layoutManager;

    public XActionRecyclerView(Context context) {
        super(context);
        layoutManager = new LinearLayoutManager(context);
        setLayoutManager(layoutManager);
        setAdapter(new XListAdapter());
        setBackgroundColor(Color.WHITE);
        this.addItemDecoration(new DividerItemDecoration(context));
    }

    public void setReverse(boolean reverse) {
        layoutManager.setReverseLayout(reverse);
    }

    public void setActionSelector(boolean enable, int normalColor, int pressColor) {
        this.actionColorEnable = enable;
        this.actionColor = normalColor;
        this.actionPress = pressColor;
    }

    public void setSelector(boolean enable, int normalColor, int pressColor) {
        this.normalColor = normalColor;
        this.pressColor = pressColor;
        this.enableItemBackColor = enable;
    }

    public void notifyDataSetChanged() {
        getAdapter().notifyDataSetChanged();
    }

    public void setItems(Collection<T> ls) {
        this.items.clear();
        this.items.addAll(ls);
        notifyDataSetChanged();
    }

    public T getItem(int position) {
        return items.get(position);
    }

    public int getItemCount() {
        return items.size();
    }

    public long getItemId(int position) {
        return position;
    }

    protected int getItemViewType(int position) {
        return 0;
    }

    protected abstract View newView(Context context, int viewType, ViewGroup parent);

    protected abstract void bindView(ListHolder holder, View itemView, int position, T item);

    protected abstract void onItemClick(ListHolder holder, View itemView, int position, T item);

    protected abstract void onItemActionClick(ListHolder holder, View itemView, int position, View actionView, int actionIndex, T item);

    protected boolean onItemLongClick(ListHolder holder, View itemView, int position, T item) {
        return false;
    }

    public class ListHolder extends ViewHolder implements OnClickListener, OnLongClickListener {
        public LayoutParams layoutParams = new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT);
        private ActionItemView actionItemView;

        public ListHolder(View itemView) {
            super(itemView);
            itemView.setOnClickListener(this);
            itemView.setOnLongClickListener(this);
            if (itemView instanceof ActionItemView) {
                actionItemView = (ActionItemView) itemView;
                for (int i = 0; i < actionItemView.getActionCount(); ++i) {
                    View v = actionItemView.getActionView(i);
                    v.setOnClickListener(this);
                }
            }
        }

        @Override
        public void onClick(View v) {
            int pos = getAdapterPosition();
            if (v == this.itemView) {
                onItemClick(this, this.itemView, pos, getItem(pos));
            } else if (null != actionItemView) {
                for (int i = 0; i < actionItemView.getActionCount(); ++i) {
                    View av = actionItemView.getActionView(i);
                    if (v == av) {
                        onItemActionClick(this, itemView, pos, av, i, getItem(pos));
                        break;
                    }
                }
            }
        }

        @Override
        public boolean onLongClick(View v) {
            int pos = getAdapterPosition();
            return onItemLongClick(this, this.itemView, pos, getItem(pos));
        }
    }

    private class XListAdapter extends Adapter<ListHolder> {

        @Override
        public ListHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View view = newView(parent.getContext(), viewType, parent);
            if (enableItemBackColor) {
                XView.view(view).backColor(normalColor, pressColor);
            }
            if (actionColorEnable && (view instanceof ActionItemView)) {
                ActionItemView vs = (ActionItemView) view;
                for (int i = 0; i < vs.getActionCount(); ++i) {
                    View v = vs.getActionView(i);
                    XView.view(v).backColor(actionColor, actionPress);
                }
            }
            return new ListHolder(view);
        }

        @Override
        public void onBindViewHolder(ListHolder holder, int position) {
            bindView(holder, holder.itemView, position, getItem(position));
            holder.itemView.setLayoutParams(holder.layoutParams);
        }

        @Override
        public int getItemCount() {
            return items.size();
        }

        @Override
        public int getItemViewType(int position) {
            return XActionRecyclerView.this.getItemViewType(position);
        }

        @Override
        public long getItemId(int position) {
            return XActionRecyclerView.this.getItemId(position);
        }
    }
}

File: yetutil/src/main/java/net/yet/ui/widget/recycler/XActionRecyclerView.kt
package net.yet.ui.widget.recycler

import android.content.Context
import android.graphics.Color
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import net.yet.theme.Colors
import net.yet.ui.ext.backColor
import java.util.*

/**
 * Created by yet on 2015/10/28.
 */
abstract class XActionRecyclerView<T>(context: Context) : RecyclerView(context) {
	private val items = ArrayList<T>()
	private var normalColor = Color.WHITE
	private var pressColor = Colors.Fade
	private var enableItemBackColor = true
	private var actionColor = Colors.TRANS
	private var actionPress = Colors.Fade
	private var actionColorEnable = true
	private val layoutManager: LinearLayoutManager

	init {
		layoutManager = LinearLayoutManager(context)
		setLayoutManager(layoutManager)
		adapter = XListAdapter()
		setBackgroundColor(Color.WHITE)
		this.addItemDecoration(DividerItemDecoration(context))
	}

	fun setReverse(reverse: Boolean) {
		layoutManager.reverseLayout = reverse
	}

	fun setActionSelector(enable: Boolean, normalColor: Int, pressColor: Int) {
		this.actionColorEnable = enable
		this.actionColor = normalColor
		this.actionPress = pressColor
	}

	fun setSelector(enable: Boolean, normalColor: Int, pressColor: Int) {
		this.normalColor = normalColor
		this.pressColor = pressColor
		this.enableItemBackColor = enable
	}

	fun notifyDataSetChanged() {
		adapter.notifyDataSetChanged()
	}

	fun setItems(ls: Collection<T>) {
		this.items.clear()
		this.items.addAll(ls)
		notifyDataSetChanged()
	}

	fun getItem(position: Int): T {
		return items[position]
	}

	val itemCount: Int
		get() = items.size

	fun getItemId(position: Int): Long {
		return position.toLong()
	}

	protected fun getItemViewType(position: Int): Int {
		return 0
	}

	protected abstract fun newView(context: Context, viewType: Int, parent: ViewGroup): View

	protected abstract fun bindView(holder: ListHolder, itemView: View, position: Int, item: T)

	protected abstract fun onItemClick(holder: ListHolder, itemView: View, position: Int, item: T)

	protected abstract fun onItemActionClick(holder: ListHolder, itemView: View, position: Int, actionView: View, actionIndex: Int, item: T)

	protected fun onItemLongClick(holder: ListHolder, itemView: View, position: Int, item: T): Boolean {
		return false
	}

	inner class ListHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener, View.OnLongClickListener {
		var layoutParams = RecyclerView.LayoutParams(RecyclerView.LayoutParams.MATCH_PARENT, RecyclerView.LayoutParams.WRAP_CONTENT)
		private var actionItemView: ActionItemView? = null

		init {
			itemView.setOnClickListener(this)
			itemView.setOnLongClickListener(this)
			if (itemView is ActionItemView) {
				actionItemView = itemView
				for (i in 0..actionItemView!!.actionCount - 1) {
					val v = actionItemView!!.getActionView(i)
					v.setOnClickListener(this)
				}
			}
		}

		override fun onClick(v: View) {
			val pos = adapterPosition
			if (v === this.itemView) {
				onItemClick(this, this.itemView, pos, getItem(pos))
			} else if (null != actionItemView) {
				for (i in 0..actionItemView!!.actionCount - 1) {
					val av = actionItemView!!.getActionView(i)
					if (v === av) {
						onItemActionClick(this, itemView, pos, av, i, getItem(pos))
						break
					}
				}
			}
		}

		override fun onLongClick(v: View): Boolean {
			val pos = adapterPosition
			return onItemLongClick(this, this.itemView, pos, getItem(pos))
		}
	}

	private inner class XListAdapter : RecyclerView.Adapter<ListHolder>() {

		override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListHolder {
			val view = newView(parent.context, viewType, parent)
			if (enableItemBackColor) {
				view.backColor(normalColor, pressColor)
			}
			if (actionColorEnable && view is ActionItemView) {
				for (i in 0..view.actionCount - 1) {
					val v = view.getActionView(i)
					v.backColor(actionColor, actionPress)
				}
			}
			return ListHolder(view)
		}

		override fun onBindViewHolder(holder: ListHolder, position: Int) {
			bindView(holder, holder.itemView, position, getItem(position))
			holder.itemView.layoutParams = holder.layoutParams
		}

		override fun getItemCount(): Int {
			return items.size
		}

		override fun getItemViewType(position: Int): Int {
			return this@XActionRecyclerView.getItemViewType(position)
		}

		override fun getItemId(position: Int): Long {
			return this@XActionRecyclerView.getItemId(position)
		}
	}
}
